src/mem/alloc.h

```diff
@@ -623,6 +623,19 @@ namespace snmalloc
             if (!l->empty())
             {
               // Send all slots to the target at the head of the list.
+              /*
+               * This is somewhat dubious: this chunk might be either a
+               * Superslab or a Mediumslab, but we access only the
+               * get_allocator() method of their common parent class,
+               * Allocslab, which reads the allocator field, which should
+               * have common offset.
+               */
+#ifdef __clang__
```

Comment:
An `AllocSlab::get()` would certainly make sense and at this particular site, we don't need any downcasting.  I'd still want to keep the assert, tho', because while I can't really imagine a C++ compiler deciding not to lay out class members for a single inheritance chain as concatenated structs, I'm sure someone will find some epsilon performance improvement by considering something else.

Non-clang objected that `Superslab` and `Mediumslab` were not "standard layout classes" thanks to their inheritance.

Reply:
Oh, so you believe it is possible for upcast actually to change the offset, i.e. 
```C++
   Superslab* p; 
   Allocslab* a = reinterpret_cast<AllocSlab*>(p);
   assert (reinterpret_cast<void*>(p) == reinterpret_cast<void*>(a)); 
```
the assertion could fail?

Reply:
I... I am just a caveman, and C++ terrifies me, as the line goes, and `reinterpret_cast` does so more than most things in this language.  Having dug around quite a bit in C++20, though, I think that that particular assertion cannot fail:

* `reinterpret_cast<Allocslab*>(p)` is defined (7.6.1.9 clause 7) to be `static_cast<Allocslab*>(static_cast<void*>(p))`.
* `static_cast<void*>(p)` is defined (7.6.1.8 clause 4; 12.4.3.1; 7.3.11 clause 2) to have the same pointer value as `p`.
* `static_cast<Allocslab*>(...)` is also defined to leave the pointer value alone (7.6.1.8 clause 13) because `Superslab` and `Allocslab` objects are not pointer interconvertable (6.8.2).
* `reinterpret_cast<void*>(x)` is `static_cast<void*>(static_cast<void*>(x))` and this leaves the pointer value unchanged by the above and the observation that the outer cast is satisfied by an empty implicit conversion sequence.

Unfortunately, I think the case at hand is subtly different: those conversions are defined to leave the pointer value unchanged but are not necessarily defined to result in sensible pointers.  In particular, I think there is risk that, despite its type, `a` is not actually left pointing at an `Allocslab`, due to the trip through `void*` induced by the `reinterpret_cast`.  Consider, instead:

```c++
Superslab *s;
Allocslab *sa = static_cast<Allocslab*>(sa);
Mediumslab *m = reinterpret_cast<Mediumslab*>(s); // same value as s, as above
Allocslab *ma = static_cast<Allocslab*>(ma);
assert (reinterpret_cast<void*>(sa) == reinterpret_cast<void*>(ma));
```

Here, the `static_cast`s are governed by (I think, eventually) 7.3.11 clause **3**, which lacks any note that the pointer value is unchanged, even if the base class is the first or sole base class.  Pointer interconvertability would save us only if our classes were standard-layout, but they're not, because `Baseslab` has mixed access control (as did `Allocslab` prior to this PR; 11.2 clauses 3.3 and 3.4).  Again, I haven't the foggiest why anyone would actually *use* this seemingly permitted freedom in the standard, but here we are.

Reply:
I'm happy for you to leave the `static_assert` in place.  But I think guarding it for clang is negating its utility. It is going to probably be some other C++ compiler that causes the issue, I am pretty certain `offsetof` is in the standard now, so this should be valid on any C++ compiler.